capsules/src/net/sixlowpan.rs

```diff
@@ -11,91 +11,267 @@
 //!
-//! ReceiveClient Trait:
-//! - receive(..): Called after an entire IPv6 packet has been reassembled.
-//!     Returns the IPv6 buffer containing the decompressed packet, as well as
-//!     additional information. Note that this function is required to return a
-//!     static buffer, as the buffer passed into this function is owned by the
-//!     RxState and not the client
+//! ```
+//!
+//! To initialize the `Sixlowpan` struct, the following methods are exposed:
+//!
+//! ```
+//! Sixlowpan::new(radio: &'a Mac<'a>,
+//!                ctx_store: C,
+//!                tx_buf: &'static mut [u8],
+//!                clock: &'a A)
+//!                -> Sixlowpan<'a, A, C>;
+//! ```
+//! The new function returns a new Sixlowpan struct. The radio argument is any
+//! object implementing the Mac trait, while the clock implements the Alarm
+//! functionality. The ctx_store argument is any object that implements the
+//! ContextStore trait.
+//!
+//! ```
+//! Sixlowpan::add_rx_state(&self, rx_state: &'a RxState<'a>);
+//! ```
+//! This method adds an RxState to the RxState pool maintained by the Sixlowpan
+//! object. Each RxState struct represents the ability to reconstruct a single
+//! IP packet at a time; if two RxState structs are allocated, two IPv6 packets
+//! can be reconstructed simultaneously (likewise for 3+ RxStates). If only
+//! a single RxState is allocated, only one IP packet can be reconstructed at
+//! a time. Note that if no RxStates are initialized, no IP packets can be
+//! received.
+//!
+//! ```
+//! Sixlowpan::set_client(&'a self, client: &'a SixlowpanClient);
+//! ```
+//! The `set_client` method sets the SixlowpanClient for the Sixlowpan object.
+//! Whatever object is set as the client receives both `receive` and `send_done`
+//! callbacks.
 //!
 //! Usage
 //! -----
 //! Examples of how to interface and use this layer are included in the file
-//! `boards/imixv1/src/lowpan_frag_dummy.rs`. Significant set-up is required
-//! in `boards/imixv1/src/main.rs` to initialize the various state for the
-//! layer and its clients.
+//! `boards/imix/src/lowpan_frag_dummy.rs`. Some set up is required in
+//! the `boards/imix/src/main.rs` file, but for the testing suite, a helper
+//! initialization function is included in the `lowpan_frag_dummy.rs` file.
+
+// Internal Design
+// ---------------
+// The overall 6LoWPAN protocol is non-trivial, and as a result, this layer
+// is fairly complex. There are two main aspects of the 6LoWPAN layer; first
+// is compression, which is abstracted as a distinct library (found at
+// `capsules/src/net/sixlowpan_compression.rs`), and second is the
+// fragmentation and reassembly layer, which is implemented in this file.
+// The documentation below describes the different components of the
+// fragmentation/reassembly functionality (for 6LoWPAN compression
+// documentation, please consult `capsules/src/net/sixlowpan_compression.rs`).
+//
```

Comment:
I think the main difference is that the compression functionality stands better on its own as a library than fragmentation, as fragmentation (especially in the reassembly/receive case) acts like a full layer. Certainly it might be a good idea to separate out the interface into a separate file for readability however.